### ISSUE TITLE
Local index / Force seems not to be working

### ISSUE THREAD

**Issue by Sineos** [Koenkk/zigbee-OTA#635]
I'm trying to update an EcoDim Dimmer, but the OTA procedure is adamantly defending the position that the current SW version is newer than the online one.

**Device:** https://www.zigbee2mqtt.io/devices/Eco-Dim.07_Eco-Dim.10.html
**OTA File:** https://github.com/Koenkk/zigbee-OTA/blob/master/images/EcoDim/ZB21S3_HZC_Dimmer1_EcoDim-Zigbee%203.0_1.01_20230908.ota
**DB Entry:**
```{
    "id": 10,
    "type": "Router",
    "ieeeAddr": "0x0c4314fffe1fa6a6",
    "nwkAddr": 35467,
    "manufId": 4714,
    "manufName": "EcoDim BV",
    "powerSource": "Mains (single phase)",
    "modelId": "EcoDim-Zigbee 3.0",
    "epList": [
        1,
        242
    ],
    "endpoints": {
        "1": {
            "profId": 260,
            "epId": 1,
            "devId": 257,
            "inClusterList": [
                0,
                3,
                4,
                5,
                6,
                8,
                2821,
                4096
            ],
            "outClusterList": [
                25
            ],
            "clusters": {
                "genBasic": {
                    "attributes": {
                        "modelId": "EcoDim-Zigbee 3.0",
                        "manufacturerName": "EcoDim BV",
                        "powerSource": 1,
                        "zclVersion": 3,
                        "appVersion": 3,
                        "stackVersion": 6,
                        "hwVersion": 1,
                        "dateCode": "20210228",
                        "swBuildId": "3.09"
                    }
                },
                "genOnOff": {
                    "attributes": {
                        "onOff": 0
                    }
                },
                "genLevelCtrl": {
                    "attributes": {
                        "currentLevel": 254
                    }
                }
            },
            "binds": [
                {
                    "cluster": 6,
                    "type": "endpoint",
                    "deviceIeeeAddress": "0x9035eafffe74e7fe",
                    "endpointID": 1
                },
                {
                    "cluster": 8,
                    "type": "endpoint",
                    "deviceIeeeAddress": "0x9035eafffe74e7fe",
                    "endpointID": 1
                }
            ],
            "configuredReportings": [
                {
                    "cluster": 6,
                    "attrId": 0,
                    "minRepIntval": 0,
                    "maxRepIntval": 65000,
                    "repChange": 1,
                    "manufacturerCode": null
                },
                {
                    "cluster": 8,
                    "attrId": 0,
                    "minRepIntval": 10,
                    "maxRepIntval": 65000,
                    "repChange": 1,
                    "manufacturerCode": null
                },
                {
                    "cluster": 6,
                    "attrId": 0,
                    "minRepIntval": 0,
                    "maxRepIntval": 65000,
                    "repChange": 1
                },
                {
                    "cluster": 8,
                    "attrId": 0,
                    "minRepIntval": 10,
                    "maxRepIntval": 65000,
                    "repChange": 1
                }
            ],
            "meta": {
            }
        },
        "242": {
            "profId": 41440,
            "epId": 242,
            "devId": 97,
            "inClusterList": [
            ],
            "outClusterList": [
                33
            ],
            "clusters": {
            },
            "binds": [
            ],
            "configuredReportings": [
            ],
            "meta": {
            }
        }
    },
    "appVersion": 3,
    "stackVersion": 6,
    "hwVersion": 1,
    "dateCode": "20210228",
    "swBuildId": "3.09",
    "zclVersion": 3,
    "interviewCompleted": true,
    "meta": {
        "configured": 332242049
    },
    "lastSeen": 1735564313030
}
```

I attempted to solve this via a local index file with the following content:
```
[
    {
        "url": "ota/ZB21S3_HZC_Dimmer1_EcoDim-Zigbee_3.0_1.01_20230908.ota",
        "force": true
    }
]
```
with having the file in the respective `ota` folder. It seems to pick up the file, as it starts complaining when I rename the file and the entries no longer match.

**Log:**
`2024-12-30 14:42:26zhc:ota: [0x0c4314fffe1fa6a6 | EcoDim-Zigbee 3.0] Firmware is newer than latest available firmware.`

**Version:**
![grafik](https://github.com/user-attachments/assets/245746c0-7711-479f-a25c-548ca558b778)

Any pointer where I'm doing wrong would be appreciated.

**Comment by Nerivec** [comment]
Do you have the `debug` logs from when you trigger the OTA check? Specifically the lines where it describes the metadata of the two firmware (imageType, etc.).

**Comment by Sineos** [comment]
Thanks for the response.
Do I need to set anything special or only log level to debug? With debug there is the following information:

```
Dec 31 18:14:35 smartvm node[1774157]: [2024-12-31 18:14:35] #033[34mdebug#033[39m: #011zh:ember:uart:ash: <--- [FRAME type=DATA]
Dec 31 18:14:35 smartvm node[1774157]: [2024-12-31 18:14:35] #033[34mdebug#033[39m: #011zh:ember:uart:ash: <--- [FRAME type=DATA ackNum=2](ackRx=2 frmTx=2)
Dec 31 18:14:35 smartvm node[1774157]: [2024-12-31 18:14:35] #033[34mdebug#033[39m: #011zh:ember:uart:ash: <--- [FRAME type=DATA ackNum=2 frmNum=4](frmRx=4) Added to rxQueue
Dec 31 18:14:35 smartvm node[1774157]: [2024-12-31 18:14:35] #033[34mdebug#033[39m: #011zh:ember:uart:ash: ---> [FRAME type=ACK frmRx=5](ackRx=2)
Dec 31 18:14:35 smartvm node[1774157]: [2024-12-31 18:14:35] #033[34mdebug#033[39m: #011zh:ember:ezsp: <=== [CBFRAME: ID=69:"INCOMING_MESSAGE_HANDLER" Seq=25 Len=49]
Dec 31 18:14:35 smartvm node[1774157]: [2024-12-31 18:14:35] #033[34mdebug#033[39m: #011zh:ember:ezsp: ezspIncomingMessageHandler(): callback called with: [type=UNICAST], [apsFrame={"profileId":260,"clusterId":25,"sourceEndpoint":1,"destinationEndpoint":1,"options":64,"groupId":0,"sequence":22}], [packetInfo:{"senderShortId":35467,"senderLongId":"0x0000000000000000","bindingIndex":255,"addressIndex":32,"lastHopLqi":132,"lastHopRssi":-78,"lastHopTimestamp":431252291}], [messageContents=011301006a122c0030650903]
Dec 31 18:14:35 smartvm node[1774157]: [2024-12-31 18:14:35] #033[34mdebug#033[39m: #011zh:controller: Received payload: clusterID=25, address=35467, groupID=0, endpoint=1, destinationEndpoint=1, wasBroadcast=false, linkQuality=132, frame={"header":{"frameControl":{"frameType":1,"manufacturerSpecific":false,"direction":0,"disableDefaultResponse":false,"reservedBits":0},"transactionSequenceNumber":19,"commandIdentifier":1},"payload":{"fieldControl":0,"manufacturerCode":4714,"imageType":44,"fileVersion":50947376},"command":{"ID":1,"response":2,"parameters":[{"name":"fieldControl","type":32},{"name":"manufacturerCode","type":33},{"name":"imageType","type":33},{"name":"fileVersion","type":35}],"name":"queryNextImageRequest"}}
Dec 31 18:14:35 smartvm node[1774157]: [2024-12-31 18:14:35] #033[34mdebug#033[39m: #011z2m: Received Zigbee message from 'DimmerKitchenEG', type 'commandQueryNextImageRequest', cluster 'genOta', data '{"fieldControl":0,"fileVersion":50947376,"imageType":44,"manufacturerCode":4714}' from endpoint 1 with groupID 0
Dec 31 18:14:35 smartvm node[1774157]: [2024-12-31 18:14:35] #033[32minfo#033[39m: #011z2m:mqtt: MQTT publish: topic 'zigbee2mqtt/DimmerKitchenEG', payload '{"last_seen":"2024-12-31T18:14:35+01:00","linkquality":132}'
Dec 31 18:14:35 smartvm node[1774157]: [2024-12-31 18:14:35] #033[34mdebug#033[39m: #011zhc:ota: [0x0c4314fffe1fa6a6 | EcoDim-Zigbee 3.0] Got request '{"fieldControl":0,"manufacturerCode":4714,"imageType":44,"fileVersion":50947376}'
Dec 31 18:14:35 smartvm node[1774157]: [2024-12-31 18:14:35] #033[34mdebug#033[39m: #011zhc:ota: [0x0c4314fffe1fa6a6 | EcoDim-Zigbee 3.0] Checking latest image availability, current: {"fieldControl":0,"manufacturerCode":4714,"imageType":44,"fileVersion":50947376}
Dec 31 18:14:35 smartvm node[1774157]: [2024-12-31 18:14:35] #033[34mdebug#033[39m: #011zhc:ota: [0x0c4314fffe1fa6a6 | EcoDim-Zigbee 3.0] Getting image metadata...
Dec 31 18:14:36 smartvm node[1774157]: [2024-12-31 18:14:36] #033[34mdebug#033[39m: #011zhc:ota: Downloaded main index
Dec 31 18:14:36 smartvm node[1774157]: [2024-12-31 18:14:36] #033[34mdebug#033[39m: #011zhc:ota: Loading override index '/opt/zigbee2mqtt/data/my_index.json'
Dec 31 18:14:36 smartvm node[1774157]: [2024-12-31 18:14:36] #033[34mdebug#033[39m: #011zhc:ota: [0x0c4314fffe1fa6a6 | EcoDim-Zigbee 3.0] Result for latest image availability, meta: '{"fileName":"Zigbee_Dimmer_1_Gang.ota","fileVersion":10,"fileSize":287022,"url":"https://github.com/Koenkk/zigbee-OTA/raw/master/images/EcoDim/Zigbee_Dimmer_1_Gang.ota","imageType":44,"manufacturerCode":4714,"sha512":"327e1afaec8baf9d367c5cd43d3dcda80e6014e56a3e400a9094249dd74e1eb2354b7106dd119b3eb661e9051cf64a2f9273a66ba0f089f7374eacb410a749bf","otaHeaderString":"EBL Zigbee_Dimmer_1_Gang\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000"}'
Dec 31 18:14:36 smartvm node[1774157]: [2024-12-31 18:14:36] #033[34mdebug#033[39m: #011zhc:ota: [0x0c4314fffe1fa6a6 | EcoDim-Zigbee 3.0] Upgrade available: NO
Dec 31 18:14:36 smartvm node[1774157]: [2024-12-31 18:14:36] #033[33mwarning#033[39m: #011zhc:ota: [0x0c4314fffe1fa6a6 | EcoDim-Zigbee 3.0] Firmware is newer than latest available firmware.
Dec 31 18:14:36 smartvm node[1774157]: [2024-12-31 18:14:36] #033[32minfo#033[39m: #011z2m: No update available for 'DimmerKitchenEG'
Dec 31 18:14:36 smartvm node[1774157]: [2024-12-31 18:14:36] #033[32minfo#033[39m: #011z2m:mqtt: MQTT publish: topic 'zigbee2mqtt/DimmerKitchenEG', payload '{"last_seen":"2024-12-31T18:14:35+01:00","linkquality":132,"update":{"installed_version":50947376,"latest_version":10,"state":"idle"}}'
Dec 31 18:14:36 smartvm node[1774157]: [2024-12-31 18:14:36] #033[32minfo#033[39m: #011z2m:mqtt: MQTT publish: topic 'zigbee2mqtt/bridge/response/device/ota_update/check', payload '{"data":{"id":"DimmerKitchenEG","update_available":false},"status":"ok","transaction":"wd2uy-2"}'
Dec 31 18:14:38 smartvm node[1774157]: [2024-12-31 18:14:38] #033[34mdebug#033[39m: #011zh:ember:uart:ash: <--- [FRAME type=DATA]
Dec 31 18:14:38 smartvm node[1774157]: [2024-12-31 18:14:38] #033[34mdebug#033[39m: #011zh:ember:uart:ash: <--- [FRAME type=DATA ackNum=2](ackRx=2 frmTx=2)
Dec 31 18:14:38 smartvm node[1774157]: [2024-12-31 18:14:38] #033[34mdebug#033[39m: #011zh:ember:uart:ash: <--- [FRAME type=DATA ackNum=2 frmNum=5](frmRx=5) Added to rxQueue
Dec 31 18:14:38 smartvm node[1774157]: [2024-12-31 18:14:38] #033[34mdebug#033[39m: #011zh:ember:uart:ash: ---> [FRAME type=ACK frmRx=6](ackRx=2)
Dec 31 18:14:38 smartvm node[1774157]: [2024-12-31 18:14:38] #033[34mdebug#033[39m: #011zh:ember:ezsp: <=== [CBFRAME: ID=69:"INCOMING_MESSAGE_HANDLER" Seq=25 Len=49]
Dec 31 18:14:38 smartvm node[1774157]: [2024-12-31 18:14:38] #033[34mdebug#033[39m: #011zh:ember:ezsp: ezspIncomingMessageHandler(): callback called with: [type=UNICAST], [apsFrame={"profileId":260,"clusterId":25,"sourceEndpoint":1,"destinationEndpoint":1,"options":64,"groupId":0,"sequence":23}], [packetInfo:{"senderShortId":35467,"senderLongId":"0x0000000000000000","bindingIndex":255,"addressIndex":32,"lastHopLqi":132,"lastHopRssi":-78,"lastHopTimestamp":434273029}], [messageContents=011401006a122c0030650903]
Dec 31 18:14:38 smartvm node[1774157]: [2024-12-31 18:14:38] #033[34mdebug#033[39m: #011zh:controller: Received payload: clusterID=25, address=35467, groupID=0, endpoint=1, destinationEndpoint=1, wasBroadcast=false, linkQuality=132, frame={"header":{"frameControl":{"frameType":1,"manufacturerSpecific":false,"direction":0,"disableDefaultResponse":false,"reservedBits":0},"transactionSequenceNumber":20,"commandIdentifier":1},"payload":{"fieldControl":0,"manufacturerCode":4714,"imageType":44,"fileVersion":50947376},"command":{"ID":1,"response":2,"parameters":[{"name":"fieldControl","type":32},{"name":"manufacturerCode","type":33},{"name":"imageType","type":33},{"name":"fileVersion","type":35}],"name":"queryNextImageRequest"}}
Dec 31 18:14:38 smartvm node[1774157]: [2024-12-31 18:14:38] #033[34mdebug#033[39m: #011z2m: Received Zigbee message from 'DimmerKitchenEG', type 'commandQueryNextImageRequest', cluster 'genOta', data '{"fieldControl":0,"fileVersion":50947376,"imageType":44,"manufacturerCode":4714}' from endpoint 1 with groupID 0
Dec 31 18:14:38 smartvm node[1774157]: [2024-12-31 18:14:38] #033[32minfo#033[39m: #011z2m:mqtt: MQTT publish: topic 'zigbee2mqtt/DimmerKitchenEG', payload '{"last_seen":"2024-12-31T18:14:38+01:00","linkquality":132}'
Dec 31 18:14:38 smartvm node[1774157]: [2024-12-31 18:14:38] #033[34mdebug#033[39m: #011z2m: Device 'DimmerKitchenEG' requested OTA
```

It seems the OTA completely disregards the file that came with https://github.com/Koenkk/zigbee-OTA/pull/357

**Comment by Nerivec** [comment]
Seems that device has multiple variants, each with their own image type (and manuf code). Yours is 44, and the one you thought it was is 1000.
```json
{"fieldControl":0,"manufacturerCode":4714,"imageType":44,"fileVersion":50947376}
```

which matches this one:
https://github.com/Koenkk/zigbee-OTA/blob/972db3ec6e8190f85052303976ffcd60d6868307/index.json#L468-L477

<details><summary>not these</summary>

https://github.com/Koenkk/zigbee-OTA/blob/972db3ec6e8190f85052303976ffcd60d6868307/index.json#L458-L467

https://github.com/Koenkk/zigbee-OTA/blob/972db3ec6e8190f85052303976ffcd60d6868307/index.json#L4154-L4163

</details>

So, you indeed have a more recent version that the one currently available in zigbee-OTA (50947376 > 10).

**Comment by Sineos** [comment]
Thanks for the insights. Appreciated.
